### ISSUE TITLE
Emulating multiple devices with a single GPU

### ISSUE THREAD

**Issue by liecn** [Lightning-AI/pytorch-lightning#19507]
### Bug description

Referring to issues #8630 and #8634, it appears that when utilizing Gloo, encountering duplicated device IDs is a possibility. However, the existing validation mechanism rejects any device list that is not entirely unique, even in scenarios where Gloo is being employed.

I have the same question mentioned in #8630 
```
Please, is there any way to split a single GPU into multiple replicas with Lightning?
Thanks!

P.S.: Ray has a really nice support for fractional GPUs: https://docs.ray.io/en/master/using-ray-with-gpus.html#fractional-gpus. I've never used them with Lightning, but maybe it could be a workaround?
```

### What version are you seeing the problem on?

v2.0

### Error messages and logs

```
# Error messages and logs here please
```

### Environment

<details>
  <summary>Current environment</summary>

```
#- Lightning Component (e.g. Trainer, LightningModule, LightningApp, LightningWork, LightningFlow):
#- PyTorch Lightning Version (e.g., 1.5.0):
#- Lightning App Version (e.g., 0.5.2):
#- PyTorch Version (e.g., 2.0):
#- Python version (e.g., 3.9):
#- OS (e.g., Linux):
#- CUDA/cuDNN version:
#- GPU models and configuration:
#- How you installed Lightning(`conda`, `pip`, source):
#- Running environment of LightningApp (e.g. local, cloud):
```

</details>

cc @borda

**Comment by awaelchli** [comment]
@liecn This is not documented and not a feature of Lightning, so I'm removing the bug label. 

To my knowledge, what you are asking is not possible in PyTorch. The link you posted does not work, but after taking a quick look at this [example](https://github.com/ray-project/ray/blob/master/rllib/examples/fractional_gpus.py) it is means that Ray is just launching multiple trials on the same device. This is not the same as using "multi-GPU" in Lightning where mutiple GPUs are working together to train.

If you want to achieve the same thing in Lightning/PyTorch as Ray does, then that is possible. You just launch two instances of the same experiment on the same GPU, like

```
CUDA_VISIBLE_DEVICES=0 python train.py --learning_rate=0.01
CUDA_VISIBLE_DEVICES=0 python train.py --learning_rate=0.02
```

**Comment by liecn** [comment]
I'm still feeling a bit puzzled about the whole setup.

Right now, I've got a setup with one node that's equipped with 4 GPUs. Unfortunately, the setup is pretty slow since I'm limited to running just one process per GPU, and on top of that, each process is only using about 1GB of memory.

So, I'm thinking about trying to launch multiple processes on a single GPU, ideally around 4. This way, I could have a total of 16 processes working together to train the same model at the same time, which I'm hoping would speed things up quite a bit.

I went ahead and disabled the check that ensures each GPU device ID is unique, allowing for duplicates in the device list. It looks like it's working now, particularly with Gloo as the backend.

**Comment by awaelchli** [comment]
> since I'm limited to running just one process per GPU

I don't see this as a limitation. This is very normal and desired. 

If you're not using all of the GPU (memory or util), then why not just increase the batch size? Increasing the batch size will always lead to an increase in throughput. This is equivalent to scaling the world size but doesn't introduce any overhead. 

For multi-GPU training, you should use the NCCL backend over GLOO for speed. So definitely consider bumping the batch size before you go with your setup.

**Comment by ksasso1028** [comment]
@awaelchli there is actually more logic to this than meets the eye. I am in a situation where the number of steps lead to better convergence. So by using a larger batch size, you actually run through much less steps for any given time frame. So I would like to do the same, and run 3 or 4 processes  on a single GPU (as if it were being distributed to other devices) for a single model during training. In this way, I can double or triple the number of steps I am running through in a given hour, thus leading to faster convergence. How can we do this?

as seen here https://github.com/Lightning-AI/pytorch-lightning/discussions/8630

The user was able to use the same GPU for multiple processes in a  _single_  training run. This is not allowed in current version since lightning complains about the same device twice. 

@liecn  @vanpelt @kashif @ozen
